src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java

```diff
@@ -0,0 +1,213 @@
+
+		// logging
+		currentPoseString = "";
+		targetPoseString = "";
+
+		networkTable = NetworkTableInstance.getDefault().getTable("limelight");
+		ShuffleboardTab limelightTab = Shuffleboard.getTab("Limelight");
+
+		setPipeline();
+
+		limelightTab.addBoolean("hasTarget", this::hasTargets).withPosition(0, 0).withSize(1, 1);
+		limelightTab
+				.addDouble("Horizontal Offset", this::getHorizontalOffset)
+				.withPosition(1, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addDouble("Vertical Offset", this::getVerticalOffset)
+				.withPosition(2, 0)
+				.withSize(1, 1);
+
+		limelightTab
+				.addDouble("Target Distance ", this::getDistanceFromTarget)
+				.withPosition(3, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addDouble("Target Distance 2 - TEST ", this::getDistanceFromTargetTheSecond)
+				.withPosition(4, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addDouble("Limelight Based Turn Power - TEST ", this::turnPowerLin)
+				.withPosition(5, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addString("Current Pose ", this::getCurrentPoseString)
+				.withPosition(0, 1)
+				.withSize(4, 1);
+
+		limelightTab
+				.addString("Target Pose ", this::getTargetPoseString)
+				.withPosition(0, 2)
+				.withSize(4, 1);
+	}
+
+	public void setPipeline() {
+		networkTable.getEntry("pipeline").setNumber(2);
+	}
+
+	// METHODS
+
+	public boolean hasTargets() {
+		return (networkTable.getEntry("tv").getDouble(0) != 0);
+	}
+
+	public double getHorizontalOffset() {
+		return networkTable.getEntry("tx").getDouble(0);
+	}
+
+	public double getVerticalOffset() {
+		return networkTable.getEntry("ty").getDouble(0);
+	}
+
+	public double getBoxWidth() {
+		return networkTable.getEntry("tshort").getDouble(0);
+	}
+
+	public double getDistanceFromTarget() {
+
+		// 1.3 at 1.9
+		// 1.0 at 1.3
+
+		double angleToTarget = getVerticalOffset();
+		// return (TARGET_HEIGHT - CAMERA_MOUNT_HEIGHT) / Math.tan(CAMERA_ANGLE_OFFSET + angleToTarget);
+
+		return (TARGET_HEIGHT - CAMERA_MOUNT_HEIGHT)
+				/ (CAMERA_ANGLE_OFFSET + Math.tan(Units.degreesToRadians(getVerticalOffset())));
+	}
+
+	public double getDistanceFromTargetTheSecond() {
+
+		// focal length = (P x D) / W
+		double focal_length = 346.1818;
+
+		// distance = (W x F) / P
+		// returns inches for testing purposes, will divide by 39.3700787 to return meters
+		return (8.25 * focal_length) / getBoxWidth();
+	}
+
+	public double turnPowerExp() {
+		if (Math.abs(getHorizontalOffset()) >= 1.0) {
+			return MathUtil.clamp(
+					(Math.signum(getHorizontalOffset())
+							* (Math.pow(Math.abs(getHorizontalOffset()), 0.6) * 0.14)),
+					-0.25,
+					0.25);
+		} else {
+			return 0.0;
+		}
+	}
+
+	public double turnPowerLin() {
+		if (Math.abs(getHorizontalOffset()) >= 1.0) {
+			return MathUtil.clamp(0.04 * getHorizontalOffset(), -0.25, 0.25);
+		} else {
+			return 0.0;
+		}
+	}
+
+	// tan(degree) * distance = sideways distance
+
+	// target height / tan(vertical angle)
+
+	// TODO fix this or something
```

Comment:
Please add more about why this is broken, and what needs fixing

Reply:
done

Reply:
(removed the TODO as it is fixed)